# A/B Test Setup
**Goal:** Randomly split users into old and new checkout, and simulate the effect of the new checkout.
**Steps:** load clean users, assign groups, simulate the new checkout's effect (bigger on mobile), save to `experiment_results`.
**Note:** The data is synthetic, so the new checkout's effect is an assumption I built in.

In [ ]:
import pandas as pd
import numpy as np
from sqlalchemy import create_engine

engine = create_engine("postgresql+psycopg2://postgres:YOUR_PASSWORD@localhost:5432/funnel_doctor")

# 1. One row per clean user: did they start checkout? did they pay?
query = """
SELECT u.user_id, u.device,
       MAX(CASE WHEN e.event = 'start_checkout'   THEN 1 ELSE 0 END) AS started_checkout,
       MAX(CASE WHEN e.event = 'payment_success'  THEN 1 ELSE 0 END) AS paid_old
FROM users_clean u
JOIN events_clean e ON e.user_id = u.user_id
GROUP BY u.user_id, u.device
"""
df = pd.read_sql(query, engine)
print("Users:", len(df))

# 2. Randomly assign old / new (seed makes it repeatable)
rng = np.random.default_rng(42)
df["variant"] = rng.choice(["old", "new"], size=len(df))

# 3. Simulate the new checkout: it helps mobile a lot, others a little
is_mobile = df["device"].str.lower() == "mobile"
chance_to_convert = np.where(is_mobile, 0.13, 0.058)

would_convert = (
    (df["variant"] == "new")
    & (df["started_checkout"] == 1)
    & (df["paid_old"] == 0)
    & (rng.random(len(df)) < chance_to_convert)
)
df["paid"] = np.where(would_convert, 1, df["paid_old"])

# 4. Save to PostgreSQL
df[["user_id", "device", "variant", "started_checkout", "paid"]].to_sql(
    "experiment_results", engine, if_exists="replace", index=False
)

# 5. Quick checks
print(df["variant"].value_counts())
checkout = df[df["started_checkout"] == 1]
print(checkout.groupby("variant")["paid"].mean())

Users: 100000
variant
new    50089
old    49911
Name: count, dtype: int64
variant
new    0.559949
old    0.517286
Name: paid, dtype: float64


**Check:** 100,000 users, split about 50/50. Payment rate after starting checkout: old 51.7%, new 56.0%.